pkg/experiment/metastore/metastore_state_poll_compaction_jobs.go

```diff
@@ -340,3 +217,79 @@ func (m *metastoreState) findJobsToAssign(jobCapacity int, raftLogIndex uint64,
 
 	return jobsToAssign
 }
+
+func (m *metastoreState) writeToDb(sTable *pollStateUpdate) error {
+	return m.db.boltdb.Update(func(tx *bbolt.Tx) error {
+		for shard, blocks := range sTable.newBlocks {
```

Comment:
We need to ensure atomicity of operations within a shard. Consider a case:
- We remove source blocks.
- We handle query that targets the blocks we just have removed.  
- We add compacted blocks.

For me it's not very apparent how we're handling this. I suspect that the query will return incomplete results

Reply:
Good point!

Currently from what I can see in the read path we read data directly from memory. 9442699 locks the shards mutex while replacing source blocks with compacted ones so things should be a bit better. Ideally we should lock a single shard but this makes the code more complex and swapping of blocks should anyway be quite fast.